### PR TITLE
test(testing): auto-delete citest W&B round-trip checkpoint artifacts

### LINKED ISSUE #1585: test(testing): live W&B round-trip tests leak full checkpoint artifacts
Follow-up to #1572 (which closed #92).

## Problem

The live `${wandb:...}` resolver round-trip tests — `test_eval.py::test_evaluate_loads_wandb_resolved_checkpoint_and_runs_inference` and `test_train.py::test_train_resumes_from_wandb_resolved_checkpoint` — publish a **full Lightning checkpoint** to the `synth-setter-citest` project via `tests/helpers/wandb_artifacts.py::publish_checkpoint_artifact`. Each run used a **fixed** artifact name (`model-citest-ffn_full-eval` / `-resume`), so every CI invocation committed a new `:latest` version that **was never deleted** — slowly accumulating against the shared 5 GB W&B storage budget that #1572 set out to protect, and risking `:latest` collisions across parallel runs.

These are the only remaining sites that upload full checkpoint bytes to W&B; production code is reference-only after #1572.

## Fix

- Make `publish_checkpoint_artifact` a context manager that names each artifact `model-citest-<id>-<token>` (unique per run) and deletes **both the artifact and its run** on exit, best-effort.
- Wrap both call sites in the `with`-block so the resolver downloads before teardown.
- Add network-free unit tests (mocked `wandb`) pinning the unique-naming and delete-on-exit contract.

## Acceptance

- No `synth-setter-citest` artifact survives a completed round-trip test.
- Parallel runs never collide on a shared `:latest`.
- A teardown failure never masks the test result.

### PR DESCRIPTION
**Pull request by ktinubu**
## Problem

The live `${wandb:...}` resolver round-trip tests — `test_eval.py::test_evaluate_loads_wandb_resolved_checkpoint_and_runs_inference` and `test_train.py::test_train_resumes_from_wandb_resolved_checkpoint` — published a full Lightning checkpoint to the `synth-setter-citest` project via `tests/helpers/wandb_artifacts.py::publish_checkpoint_artifact` under a **fixed** artifact name. Every CI run committed a new `:latest` version that was **never deleted**, slowly accumulating against the shared 5 GB W&B storage budget that #1572 set out to protect, and risking `:latest` collisions across parallel runs.

These were the only remaining sites uploading full checkpoint *bytes* to W&B; production code is reference-only after #1572.

## Change

- `publish_checkpoint_artifact` is now a context manager that names each artifact `model-citest-<id>-<token>` (unique per run, so parallel CI never collides on `:latest`) and deletes **both the artifact and its run** on context exit — best-effort, so a teardown failure never masks the test result.
- Both call sites run inside the `with` block so the resolver downloads before teardown.
- Adds network-free unit tests (mocked `wandb` SDK) pinning unique-naming, delete-on-exit, delete-on-exception, and swallowed-cleanup-failure.

## Verification

- New unit tests: 4 passed. `ruff` / `ruff format` / `pyright` / `pydoclint` / full pre-commit on the changed files: pass.
- **Live W&B round-trip:** `test_train_resumes_from_wandb_resolved_checkpoint[cpu]` passed (77s). A direct probe confirmed the artifact resolves **inside** the context (`COMMITTED`) and `Api().artifact()` raises `CommError` **after** exit — real deletion verified.
- Pre-PR `/repo-review-full-no-comments`: 0 BLOCK across 5 skills; the actionable WARNs were addressed.

Fixes #1585
Refs #92

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

The promise: the citest round-trip **artifact and its run are deleted on context exit**, names are **unique per run**, cleanup is **best-effort** (never masks the test result), and the resolver round-trip still works through the new context manager.

- [x] **Live round-trip: both the artifact and its run are deleted on context exit** *(Level 3 — triggers real W&B with real input; this is the core promise. The e2e test below asserts resume, not deletion, so deletion is proven here directly.)*
  ```bash
  $ uv run python - < /tmp/verify_pr1590_probe.py   # publishes via the helper, inspects W&B inside vs. after the context
  INSIDE_artifact_state: COMMITTED
  INSIDE_run: khaledtinubu-n-a/synth-setter-citest/nk5e7x3d
  ARTIFACT_AFTER_EXIT: DELETED via CommError
  RUN_AFTER_EXIT: DELETED via CommError
  UNIQUE_NAMES_DISTINCT: True | ['model-citest-verify-1617df5c:latest', 'model-citest-verify-a74eafaf:latest']
  ```
  Inside the `with`, `Api().artifact(ref)` resolves (`COMMITTED`); after exit, both `Api().artifact(ref)` and `Api().run(run_path)` raise `CommError` — the artifact **and** its run are gone. (Probe script: publishes a real checkpoint, reads `logged_by()` for the run path inside the context, re-queries after.)

- [x] **Unique per-run artifact names — no `:latest` collision across concurrent runs** *(Level 3 — adversarial: the prior fixed-name scheme is exactly what let CI runs pile up versions; two live publishes of the same base name must differ.)*
  Last line above: two publishes of base `model-citest-verify` yielded `…-1617df5c` and `…-a74eafaf` (`UNIQUE_NAMES_DISTINCT: True`).

- [x] **End-to-end resolver resume still works through the `with`-wrapped helper** *(Level 1 — promise is the round-trip is preserved: full train → publish → `${wandb:…}` resolve → download → resume → teardown.)*
  ```bash
  $ uv run pytest "tests/test_train.py::test_train_resumes_from_wandb_resolved_checkpoint[cpu-surge/ffn_full]"
  1 passed, 2 warnings in 60.35s (0:01:00)
  ```
  Real key + VST present (`WANDB_API_KEY`, `SYNTH_SETTER_PLUGIN_PATH` set); the test trains a real checkpoint, publishes it, resolves it back via `${wandb:…}`, and asserts `global_step` advances — i.e. the resolver downloaded before teardown deleted the artifact.

- [x] **Cleanup failure-modes: delete-on-exception and swallowed-cleanup-failure** *(Level 2 — these require forcing the body to raise and forcing the W&B delete to raise; that can only be exercised with a stubbed SDK. The happy path is covered live above.)*
  ```bash
  $ uv run pytest tests/test_wandb_artifacts_helper.py -v
  test_publish_checkpoint_artifact_deletes_artifact_and_run_on_exit PASSED
  test_publish_checkpoint_artifact_deletes_even_when_body_raises PASSED
  test_publish_checkpoint_artifact_swallows_cleanup_failure PASSED
  test_publish_checkpoint_artifact_embeds_ckpt_and_yields_suffixed_latest_ref PASSED
  4 passed in 0.34s
  ```
  `…_deletes_even_when_body_raises` pins that teardown runs when the `with` body raises; `…_swallows_cleanup_failure` pins that a W&B delete error during teardown does not propagate (and `run.finish()` + the run-delete are still attempted).

**Probe script** (`/tmp/verify_pr1590_probe.py`, run from the worktree so `tests/` is the import root):
```python
import tempfile, pathlib, torch
from tests.helpers.wandb_artifacts import publish_checkpoint_artifact
import wandb

tmp = pathlib.Path(tempfile.mkdtemp())
ckpt = tmp / "model.ckpt"
torch.save({"state_dict": {}}, ckpt)

names = []
run_path = None
with publish_checkpoint_artifact(ckpt, "model-citest-verify", tmp / "wb1") as ref:
    art = wandb.Api().artifact(ref)
    print("INSIDE_artifact_state:", art.state)
    r = art.logged_by()
    run_path = f"{r.entity}/{r.project}/{r.id}"
    print("INSIDE_run:", run_path)
    names.append(ref.split("/")[-1])

try:
    wandb.Api().artifact(ref)
    print("ARTIFACT_AFTER_EXIT: STILL_EXISTS_FAIL")
except Exception as e:
    print("ARTIFACT_AFTER_EXIT: DELETED via", type(e).__name__)

try:
    wandb.Api().run(run_path)
    print("RUN_AFTER_EXIT: STILL_EXISTS_FAIL")
except Exception as e:
    print("RUN_AFTER_EXIT: DELETED via", type(e).__name__)

with publish_checkpoint_artifact(ckpt, "model-citest-verify", tmp / "wb2") as ref2:
    names.append(ref2.split("/")[-1])
print("UNIQUE_NAMES_DISTINCT:", names[0] != names[1], "|", names)
```

File: tests/helpers/wandb_artifacts.py [hunk 1]
```diff
@@ -42,13 +54,35 @@ def publish_checkpoint_artifact(ckpt_path: Path, artifact_name: str, run_dir: Pa
...
         artifact.add_file(str(ckpt_path), name="model.ckpt")
         run.log_artifact(artifact)
         artifact.wait()
-        # Read the entity back from the run so the ref matches whatever account the key owns.
-        ref = f"{run.entity}/{CITEST_PROJECT}/{artifact_name}:latest"
+        yield f"{entity}/{CITEST_PROJECT}/{unique_name}:latest"
     finally:
         run.finish()
-    return ref
+        _delete_citest_artifact_and_run(entity, unique_name, run_id)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/helpers/wandb_artifacts.py:70]: Good catch — fixed in 9b18abc3. `run.finish()` is now wrapped in `contextlib.suppress(Exception)` inside the `finally`, so a transient comm error there neither fails the test nor short-circuits the artifact/run deletion. Added `test_publish_checkpoint_artifact_finish_failure_does_not_skip_deletion` to pin that a raising `finish()` still reaches both deletions without escaping the context.
